check-submodules.sh

```diff
@@ -0,0 +1,146 @@
+#!/bin/bash
+set -euo pipefail
+IFS=$'\n\t'
+
+SCRIPT_ROOT="$(cd -P "$( dirname "$0" )" && pwd)"
+SUBMODULES=$SCRIPT_ROOT/.gitmodules
+
+# has the user answered "yes to all" to init'ing submodules?
+answered_all=0
+
+# ask for confirmation and initialize a submodule if approved.
+init_submodule () {
+  module=$1
+  done=0
+  while [ $done == 0 ]; do
+    echo "warning: submodules $module does not appear to be initialized."
+    read -p "Should I initialize it for you? [Yes/no/all/quit] " -n 1 -r
+    echo
+    if [[ $REPLY =~ ^[Aa]$ ]]; then
+      answered_all=1
+      done=1
+      git submodule update --init --recursive $module
+    elif [[ $REPLY =~ ^[Qq]$ ]]; then
+      exit 0
+    elif [[ $REPLY =~ ^[Nn]$ ]]; then
+      done=1
+    elif [[ $REPLY == "" || $REPLY == " " || $REPLY =~ ^[Yy]$ ]]; then
+      done=1
+      git submodule update --init --recursive $module
+    else
+      echo "didn't understand that ($REPLY)"
+    fi
+  done
+}
+
+# update a submodule to an expected commit.  We give different messages
+# for being ahead vs being behind or diverged from the expected commit,
+# so this function is parameterized.
+fix_submodule () {
+  path=$1
+  expected=$2
+  actual=$3
+  msg=$4
+  prompt=$5
+  done=0
+  while [ $done == 0 ]; do
+    echo -e "$msg"
+    read -p "$prompt " -n 1 -r
+    echo
+    if [[ $REPLY =~ ^[Qq]$ ]]; then
+      exit 1
+    elif [[ $REPLY =~ ^[Yy]$ ]]; then
+	  # check if we have this commit locally and can skip the fetch
+      git cat-file -e $expected^{commit} 2>/dev/null || exitCode=$?
+      if [ $exitCode != 0 ]; then
+        git fetch
+      fi
+      exitCode=0
+	  # double-check, we should have the commit now unless something
+	  # weird is going on.
```

Comment:
Tab chars before the comments in this function are causing alignment issues.

Reply:
Fixed.